### PR TITLE
Support #if SNIPPET in snippets

### PR DESCRIPTION
**Pull request by pakrym**
It's closer to C# syntax, can be used to verify that snippets still build after replacement and I can type it without having to consult the doc.

### CODE REVIEW

**Comment by pakrym** [comment]
If we decide to do this I'll update all existing snippets and remove the commend syntax.

File: eng/SnippetGenerator/DirectoryProcessor.cs [hunk 1]
```diff
@@ -225,9 +226,13 @@ private List<Snippet> GetSnippetsInDirectory(string baseDirectory)
         private Snippet[] GetAllSnippets(SyntaxTree syntaxTree)
         {
             var snippets = new List<Snippet>();
+            var newRoot = PreprocessorDirectiveRemover.Instance.Visit(syntaxTree.GetRoot());
+            syntaxTree = syntaxTree.WithRootAndOptions(newRoot, syntaxTree.Options);
+
             var directiveWalker = new DirectiveWalker();
             directiveWalker.Visit(syntaxTree.GetRoot());
 
+
```
Comment (christothes) [inline eng/SnippetGenerator/DirectoryProcessor.cs:235]: nit: extra line

**Review by christothes — APPROVED** [review]
Nice!

File: sdk/monitor/Azure.Monitory.Query/tests/LogsClientSamples.cs [hunk 2]
```diff
@@ -74,8 +77,11 @@ public async Task QueryLogsAsPrimitive()
             #region Snippet:QueryLogsAsPrimitive
 
             LogsClient client = new LogsClient(new DefaultAzureCredential());
-            /*@@*/string workspaceId = TestEnvironment.WorkspaceId;
-            //@@string workspaceId = "<workspace_id>";
+#if SNIPPET
+            string workspaceId = "<workspace_id>";
+#else
```
Comment (heaths) [inline sdk/monitor/Azure.Monitory.Query/tests/LogsClientSamples.cs:82]: FWIW, in Key Vault there's many places we do something like this:

```csharp
KeyClient client = new KeyClient(...);
/*@@*/ client = GetClient(); // For recorded samples/tests
```

Which should work simply as:

```csharp
KeyClient client = new KeyClient(...);
#if !SNIPPET
client = GetcClient();
#endif
```

I suppose you could declare the local like this in either case as well.

**Review by heaths — APPROVED** [review]
I like this. It should also help the occasional reformatting problems where VS wants to put `/*@@*/` on it's own line and, if not caught, screws up the testable sample/snippet. At some point recently, too, VS seems to have stopped graying out ifdef'd out code and still provides IntelliSense, which helps writing syntactically correct snippet-only code.

**Comment by maririos** [comment]
@tg-msft once opened this issue => https://github.com/Azure/azure-sdk-for-net/issues/8539 
1y 1/2 later it is getting done!!

**Comment by pakrym** [comment]
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/8539

**Comment by pakrym** [comment]
>If we decide to do this I'll update all existing snippets and remove the commend syntax.

Maybe not, there are a LOT of them. I might ask service owners to clean up their samples.

**Comment by maririos** [comment]
>Maybe not, there are a LOT of them. I might ask service owners to clean up their samples.

Feel free to create issues per library and we can track the progress in the Eng Excellence epic.

**Review by jsquire — APPROVED** [review]

**Review by tg-msft — APPROVED** [review]
This is great.  A couple of questions:
- Will refactoring work in the `#if SNIPPET` branch?
- Should we file a work item to schedule a weekly build with `-d SNIPPET` to make sure all the code that doesn't run still compiles?

**Comment by pakrym** [comment]
>Will refactoring work in the #if SNIPPET branch?
No, unfortunately, Roslyn treats the text in disabled branches as absent from the syntax/semantic tree.
>Should we file a work item to schedule a weekly build with -d SNIPPET to make sure all the code that doesn't run still compiles?
Will do.

**Comment by pakrym** [comment]
Filed: https://github.com/Azure/azure-sdk-for-net/issues/20431 https://github.com/Azure/azure-sdk-for-net/issues/20432

**Pull request merged**